Update quickstarts, install, and CLI reference (#41)
agentregistry-dev/website: content/docs/operations/database.md

---
title: BYO PostgreSQL database
weight: 10
description: "Replace the built-in PostgreSQL database with an external instance for production use."
---

By default, agentregistry deploys a bundled PostgreSQL database. Because it runs as a single pod or container, data is lost when it restarts or becomes unavailable.

For production environments, replace the bundled database with an external PostgreSQL instance that you manage and back up independently.

## Docker

If you installed agentregistry with Docker Compose, edit the compose file to remove the bundled PostgreSQL service and point the server at your own database.

### Before you begin

- agentregistry installed with Docker Compose. See the [Install with Docker](/docs/install/docker) guide.

> - An external PostgreSQL instance (version 14 or later) reachable from your Docker host.

Reviewer: I think these need to be reworded into steps or needs a line before these

### Step 1: Edit the compose file

Open `agentregistry-compose.yml` and make the following changes:

1. Remove the `postgres` service block entirely.
2. Remove the `postgres_data` volume.
3. Update `AGENT_REGISTRY_DATABASE_URL` in the `agentregistry` service to point at your external database.
4. Remove the `depends_on` condition that references the bundled postgres service.

> The updated `agentregistry` service should look like this:

Reviewer: ```suggestion
The updated `agentregistry` service looks like the following example:
```

        fi
        exec /app/bin/arctl-server
    environment:
      AGENT_REGISTRY_DATABASE_URL: "postgres://user:password@your-pg-host:5432/agentregistry?sslmode=require"
      AGENT_REGISTRY_SERVER_ADDRESS: ":8080"
      AGENT_REGISTRY_ENABLE_REGISTRY_VALIDATION: "false"
      AGENT_REGISTRY_MCP_PORT: "31313"
      KUBECONFIG: "/root/.kube/config"
    ports:
      - "12121:8080"
      - "31313:31313"
    extra_hosts:
      - "host.docker.internal:host-gateway"
    volumes:
      - ~/.kube/config:/root/.kube/config.orig:ro
    healthcheck:
      test: ["CMD", "curl", "--fail", "--silent", "--show-error", "http://localhost:8080/v0/ping"]
      interval: 5s
      timeout: 2s
      retries: 12
      start_period: 5s
    networks:
      - agentregistry-network
    restart: unless-stopped

networks:
  agentregistry-network:
    driver: bridge
```

### Step 2: Create the agentregistry database

If the `agentregistry` database does not already exist on your PostgreSQL instance, create it:

```sh
psql "host=<your-pg-host> port=5432 user=<your-pg-user> dbname=postgres sslmode=require" \
  -c 'CREATE DATABASE agentregistry;'
```

### Step 3: Restart the stack

> 

Reviewer: needs some kind of step wording

```sh
docker compose -f agentregistry-compose.yml down
docker compose -f agentregistry-compose.yml up -d --wait
```

---

## Kubernetes

> If you installed agentregistry with Helm, the instructions below show how to provision an [Amazon RDS for PostgreSQL](https://aws.amazon.com/rds/postgresql/) instance in the same VPC as your EKS cluster and configure agentregistry to use it.

Reviewer: ```suggestion
If you installed agentregistry with Helm, the instructions in this section show how to provision an [Amazon RDS for PostgreSQL](https://aws.amazon.com/rds/postgresql/) instance in the same VPC as your EKS cluster and configure agentregistry to use it.
```

### Before you begin

> 

Reviewer: same

- agentregistry installed in Kubernetes via Helm. See the [Install on Kubernetes](/docs/install/kubernetes) guide.
- The [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html) configured with access to your AWS account.
- `kubectl` access to your EKS cluster.

> Set your cluster name and AWS region as environment variables:

Reviewer: these all need to be reformatted into steps. cant tell if these are meant to be in before you begin or their own section

   echo "RDS security group: $RDS_SG_ID"
   ```

2. Add an ingress rule to allow TCP traffic on port 5432 from the EKS node security group.
   ```sh
   aws ec2 authorize-security-group-ingress \
     --group-id "$RDS_SG_ID" \
     --protocol tcp \
     --port 5432 \
     --source-group "$EKS_NODE_SG" \
     --region "$AWS_REGION"
   ```

3. Retrieve the subnet IDs that your EKS cluster uses. Using the cluster's own subnets guarantees the DB subnet group covers at least two Availability Zones, which RDS requires.
   ```sh
   SUBNET_IDS=$(aws eks describe-cluster \
     --name "$EKS_CLUSTER_NAME" \
     --region "$AWS_REGION" \
     --query "cluster.resourcesVpcConfig.subnetIds" \
     --output text | tr '\t' ',')
   echo "Subnet IDs: $SUBNET_IDS"
   ```

4. Create a database subnet group. If you already have one for this VPC, skip this step and set `DB_SUBNET_GROUP` to its name.
   ```sh
   DB_SUBNET_GROUP=agentregistry-db-subnet-group

   aws rds create-db-subnet-group \
     --db-subnet-group-name "$DB_SUBNET_GROUP" \
     --db-subnet-group-description "Subnet group for agentregistry RDS" \
     --subnet-ids $(echo $SUBNET_IDS | tr ',' ' ') \
     --region "$AWS_REGION"
   ```

> 5. Create the RDS for PostgreSQL instance.

Reviewer: ```suggestion
5. Create the RDS for the PostgreSQL instance.
```

     --master-username "$PG_USER" \
     --master-user-password "$PG_PASS" \
     --db-name postgres \
     --allocated-storage 20 \
     --storage-type gp3 \
     --no-publicly-accessible \
     --vpc-security-group-ids "$RDS_SG_ID" \
     --db-subnet-group-name "$DB_SUBNET_GROUP" \
     --backup-retention-period 7 \
     --region "$AWS_REGION"
   ```

   > [!NOTE]
   > The initial database is named `postgres`. You create the `agentregistry` application database in the next step. RDS does not allow the master database to be renamed during instance creation.

6. Wait for the instance to reach the `available` state. This can take 5–10 minutes.
   ```sh
   aws rds wait db-instance-available \
     --db-instance-identifier "$DB_INSTANCE_ID" \
     --region "$AWS_REGION"
   echo "RDS instance is available."
   ```

7. Get the RDS endpoint.
   ```sh
   PG_HOST=$(aws rds describe-db-instances \
     --db-instance-identifier "$DB_INSTANCE_ID" \
     --region "$AWS_REGION" \
     --query "DBInstances[0].Endpoint.Address" \
     --output text)
   PG_PORT=5432
   echo "RDS endpoint: $PG_HOST:$PG_PORT"
   ```

### Step 2: Create the agentregistry database

> Run a temporary pod in your cluster to connect to the RDS instance and create the application database.

Reviewer: all in this section needs to be formatted into standard steps

  --namespace agentregistry \
  --reuse-values \
  --set database.postgres.type=external \
  --set database.postgres.external.secretRef.name=agentregistry-db-creds \
  --set database.postgres.external.secretRef.key=AGENT_REGISTRY_DATABASE_URL \
  --wait --timeout=5m
```

{{% /tab %}}
{{% tab name="Inline connection string" %}}

```sh
helm upgrade -i agentregistry \
  oci://ghcr.io/agentregistry-dev/agentregistry/charts/agentregistry \
  --namespace agentregistry \
  --reuse-values \
  --set database.postgres.type=external \
  --set "database.postgres.external.url=${DSN}" \
  --wait --timeout=5m
```

{{% /tab %}}
{{< /tabs >}}

| Helm value | Description |
|---|---|
| `database.postgres.type` | Set to `external` to use your own database instead of the bundled instance. |
| `database.postgres.external.url` | Inline connection string. Mutually exclusive with `secretRef.name`. |
| `database.postgres.external.secretRef.name` | Name of a Kubernetes Secret holding the connection string. Mutually exclusive with `url`. |
| `database.postgres.external.secretRef.key` | Key within the Secret (default: `AGENT_REGISTRY_DATABASE_URL`). |

### Step 5: Verify the connection

> 1. Verify the pods are healthy and that no bundled `agentregistry-postgresql-*` pod is running.

Reviewer: ```suggestion
1. Verify that the pods are healthy and that no bundled `agentregistry-postgresql-*` pod is running.
```

   ```sh
   kubectl get pods -n agentregistry
   ```

2. Confirm the server pod has the Secret reference in its environment.
   ```sh
   kubectl -n agentregistry describe pod -l app.kubernetes.io/component=server \
     | sed -n '/AGENT_REGISTRY_DATABASE_URL/,/^$/p' | head -5
   ```

>    Expected output:

Reviewer: ```suggestion
   Example output:
```

   ```console
   AGENT_REGISTRY_DATABASE_URL:  <set to the key 'AGENT_REGISTRY_DATABASE_URL' in secret 'agentregistry-db-creds'>  Optional: false
   ```

3. Check the server logs for successful migration messages.
   ```sh
   kubectl -n agentregistry logs deploy/agentregistry-server --tail=50 | grep "migration"
   ```

>    Expected output:

Reviewer: ```suggestion
   Example output:
```

   ```console
   {"time":"...","level":"info","msg":"all migrations applied successfully","component":"database.migrate"}
   ```

4. List the tables that the registry created on the RDS instance.
   ```sh
   kubectl -n agentregistry run pg-check --rm -it --image=postgres:16 --restart=Never -- \
     psql "host=$PG_HOST port=$PG_PORT user=$PG_USER password='$PG_PASS' dbname=agentregistry sslmode=require" \
     -c '\dt'

>    ```

Reviewer: maybe an example output? idk what they need to look for in the tables
